File: Problem-Set/Graph_Algorithms/KNIGHTMV/q37_mem.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<string.h>

int main()
{
    char array[10000],c;
    long int t;
    
    scanf("%ld",&t);
    scanf("%c",&c);
    
    while(t)
    {
          gets(array);
          if(((array[0]>='a') && (array[0]<='h')) && ((array[1]>='1') && (array[1]<='8')) && (array[2]=='-') && ((array[3]>='a') && (array[3]<='h')) && ((array[4]>='1') && (array[4]<='8')) && strlen(array)==5)

          {
                          
                              if((abs(array[0]-array[3])==1 && abs(array[1]-array[4])==2)||(abs(array[0]-array[3])==2 && abs(array[1]-array[4])==1))
                              {
                                 
                                     printf("Yes\n");
                                     
                              }
                                 else
                                 {
                                     printf("No\n");
                                     
                                 }
                              
      }
      else
      {
       printf("Error\n");
      }
      
      
   t--;
   
   
}

return 0;

}                          
                                          
                                                                                                                     

File: Problem-Set/Graph_Algorithms/KNIGHTMV/q37_mem.hs
import System.IO
import Data.Char
import Control.Monad
main = readLn >>= flip replicateM_ (getLine >>= putStrLn . correctMove)
correctMove :: String -> String
correctMove (sf : sr : '-' : ef : er : []) = if (sf <= 'h' && ef <= 'h' &&
	sf >= 'a' && ef >= 'a' && sr >= '1' && sr <= '8' && er >= '1' && er <= '8')
		then if (legalMove (ord sf) (ord sr) (ord ef) (ord er))
							then "Yes" else "No"
		else "Error"
correctMove _ = "Error"
legalMove :: Int -> Int -> Int -> Int -> Bool
legalMove sf sr ef er = if (abs (er - sr) == 1 && abs (ef - sf) == 2) ||
							(abs (er - sr) == 2 && abs (ef - sf) == 1)
			then True else False
